core/templates/components/skill-selector/skill-selector.component.ts

```diff
@@ -0,0 +1,213 @@
+import { downgradeComponent } from '@angular/upgrade/static';
+import { GroupedSkillSummaries } from 'pages/skill-editor-page/services/skill-editor-state.service';
+import { FilterForMatchingSubstringPipe } from 'filters/string-utility-filters/filter-for-matching-substring.pipe';
+import { ShortSkillSummary } from 'core/templates/domain/skill/ShortSkillSummaryObjectFactory';
+
+require('domain/utilities/url-interpolation.service.ts');
+
+@Component({
+  selector: 'skill-selector',
+  templateUrl: './skill-selector.component.html',
+})
+export class SkillSelectorComponent implements OnInit {
+  // If countOfSkillsToPrioritize > 0, then sortedSkillSummaries should
+  // have the initial 'countOfSkillsToPrioritize' entries of skills with
+  // the same priority.
+  @Input() sortedSkillSummaries: GroupedSkillSummaries;
+  @Input() selectedSkillId: string;
+  @Input() countOfSkillsToPrioritize: number;
+  @Input() categorizedSkills: CategorizedSkills;
+  @Input() untriagedSkillSummaries: SkillSummary[];
+  @Input() allowSkillsFromOtherTopics: boolean;
+  @Output() selectedSkillIdChanged: EventEmitter<string> = new EventEmitter();
+  currCategorizedSkills: CategorizedSkills = null;
+  selectedSkill = null;
+  skillFilterText = '';
+  topicFilterList = [];
+  subTopicFilterDict: { [topicName: string]:
+     { subTopicName: string; checked: boolean }[] } = {};
+  intialSubTopicFilterDict = {};
+
+  private filterForMatchingSubtringPipe: FilterForMatchingSubstringPipe = (
+    new FilterForMatchingSubstringPipe()
+  );
+
+  constructor() {}
+
+  ngOnInit(): void {
+    this.currCategorizedSkills = this.categorizedSkills;
+    for (let topicName in this.currCategorizedSkills) {
+      var topicNameDict = {
+        topicName: topicName,
+        checked: false
+      };
+      this.topicFilterList.push(topicNameDict);
+      var subTopics = this.currCategorizedSkills[topicName];
+      this.subTopicFilterDict[topicName] = [];
+      for (let subTopic in subTopics) {
+        var subTopicNameDict = {
+          subTopicName: subTopic,
+          checked: false
+        };
+        this.subTopicFilterDict[topicName].push(subTopicNameDict);
+      }
+    }
+    this.intialSubTopicFilterDict = angular.copy(this.subTopicFilterDict);
+  }
+
+  checkIfEmpty(skills: string[]): boolean {
+    return skills.length === 0;
+  }
+
+  checkTopicIsNotEmpty(topicName: string): boolean {
+    for (let key in this.currCategorizedSkills[topicName]) {
+      if (Object.keys(this.currCategorizedSkills[topicName][key]).length) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  setSelectedSkillId(): void {
+    this.selectedSkillIdChanged.emit(this.selectedSkill);
+  }
+
+  // The folowing function is called when the subtopic filter changes.
+  // This updates the list of Skills displayed in the selector.
+  updateSkillsListOnSubtopicFilterChange(): void {
+    var updatedSkillsDict = {};
+    var isAnySubTopicChecked = false;
```

Comment:
use const/let here. And assign a type to both variables.

Reply:
Done